pydantic/_internal/_generate_schema.py

```diff
@@ -1734,6 +1712,16 @@ def _dataclass_schema(
             if maybe_schema is not None:
                 return maybe_schema
 
+            schema = dataclass.__dict__.get('__pydantic_core_schema__')
+            if schema is not None and not isinstance(schema, MockCoreSchema):
+                schema = self._unpack_refs_defs(schema)
+                ref = get_ref(schema)
+                if ref:
+                    self.defs.definitions[ref] = schema
+                    return core_schema.definition_reference_schema(ref)
+                else:
+                    return schema
```

Comment:
Let's consolidate this shared pattern to avoid potential inconsistencies across models, dataclasses, etc

Reply:
This certainly should be in a shared pattern, but seems like https://github.com/pydantic/pydantic/pull/10655 does something similar already. Maybe it's best to have it implemented there when rebasing?

Reply:
Hmm I'd rather have it shared now, then that refactor should look more simple in the long run? That PR is already big enough 😅